site/docs/develop/kernel.md

```diff
@@ -1,47 +1,268 @@
-of the number of 4kb objects that Nessie will be storing in the objects table. Scans 
-are not done on this table except in the case of rare Garbage Collection operations. 
+Contention by itself is not avoidable, because operations against Nessie are guaranteed to be
+atomic and consistent.
 
+### Nessie Content-Types
 
-Space Consumption: The data stored in DynamoDB is designed to largely fit within DynamoDBs 4kb read size 
-unit (and 4x the write size unit). This true for objects in both the refs table and 
-the objects table.
+The state of so called `Contents` objects like `IcebergTable` or `DeltaLakeTable` represent the
+current state of a table in a data lake. Whenever a table has changed via for example Iceberg,
+a so-called commit operation instructs Nessie to record the new state in a Nessie-commit, which
+carries the `Contents` object(s).
 
-At a high level, Nessie's commit kernel breaks keyspace into 
-a 3 level tree and each operation restates some portion of that tree. Each level of 
-the tree is identified by sha256 20 byte hash value, which is used to as a storage 
-key in DynamoDB. Commit operations are atomic and interact with branch objects using 
-a data structure akin to a 151-way striped locked. This data structure is updated using 
-patch conditional operations against DynamoDB. To maintain linear history, the 
-branch object also maintains a commit log that works with the striped lock to make 
-history clear and consistent. 
+`IcebergTable` contains the _current_ and _global_ pointer to Iceberg's table-metadata plus the
+ID of the snapshot defined in the table metadata. Since Iceberg's table-metadata manages information
+that must be consistent across all branches in Nessie, it is stored as so-called _global state_.
+The value of the snapshot-ID is stored per Nessie named-reference (branch or tag).
 
+Updating _global-state_ and _on-reference-state_ are technically operations against two different
+entities in Nessie's backend database. Classic, relational databases (usually) come with a
+transaction manager, which ensures that changes to different tables appear atomically to other
+users. Much more scalable key-value stores do not have a transaction-manager, but usually only
+provide so-called "Compare-and-Swap" (CAS) operations, which conditionally update a single key-value
+pair. This means, that the data-model has to be fundamentally different for non-transactional
+key-value stores and transactional databases. Support for non-transactional databases, the
+data-model, is designed in a way that only requires a single CAS-operation to ensure atomicity and
+consistency even when committing two logical entities, namely the _global-state_ and the
+_on-reference-state_, respectively the update to the "HEAD" of the updated branch. Some more details
+are outlined below.
+
+### Version-Store and Database-Adapters
+
+Nessie's REST-API implementation works against the `VersionStore` interface, which defines the
```

Comment:
tangential: Do we need `VersionStore` now that we have `DatabaseAdapter`? Do we have `VersionStore` implementations that are not based on a `DatabaseAdapter`?

Reply:
VS knows about the concrete types. DataabseAdapter does not.